In [3]:
print(0.3-0.2 == 0.1)

print(0.3-0.2)

False
0.09999999999999998


# Why is `0.3 - 0.2 == 0.1` False in Python?

In Python (and most programming languages), `0.3 - 0.2 == 0.1` evaluates to `False` due to **floating-point precision errors**. Floating-point numbers are represented in a binary format, which cannot exactly represent some decimal fractions, leading to slight inaccuracies.

---

## Explanation:

### 1. **Binary Representation**:
- Numbers like `0.1`, `0.2`, and `0.3` are repeating fractions in binary (similar to how `1/3` is repeating in decimal).
- For example:
  - `0.1` in binary: `0.0001100110011...` (repeating).
  - These numbers are approximated to the closest value that can be stored in a fixed number of bits.

---

### 2. **What Happens in Python**:
- When you calculate `0.3 - 0.2`, Python stores an approximation, which is slightly different from `0.1`:
  - `0.3 - 0.2` results in something like `0.09999999999999998`.
  - `0.1` is stored as something like `0.10000000000000001`.
- These two values are not exactly equal, so `0.3 - 0.2 == 0.1` returns `False`.

---

## How to Handle This:

To compare floating-point numbers, use the `math.isclose` method or set a tolerance for the comparison.

### Example:
```python
import math

print(math.isclose(0.3 - 0.2, 0.1))  # True
```

Here, `math.isclose` checks if the two numbers are nearly equal within a specified tolerance (default is very small).

0.09999999999999998